### ISSUE TITLE
Shortcode handlers should respond to context; inline vs block vs code

### ISSUE THREAD

**Issue by tverbeiren** [quarto-dev/quarto-cli#5110]
### Bug description

Using [Quarto variables](https://quarto.org/docs/authoring/variables.html) makes a lot of sense when dealing with customisations. Variables can include markdown, but in our experience it's limited to hyperlinks and some basic formatting. Titles are not taken into account, code blocks are rendered as lines of verbatim code, bullet points are rendered as a long list of options.

I generated a sample project with different types of markdown content to be included in the following repo: <https://github.com/tverbeiren/markdown_variables>. Further info is available in the `index.qmd` and the `README.md` that is generated from it.

A number of attempts have been made already with different YAML multiline options, escaping of characters, adding extra newlines. Without success.

This has been tested with Quarto v1.3.310 on Mac Ventura.

### Checklist

- [X] Please include a minimal, fully reproducible example in a single .qmd file? Please provide the whole file rather than the snippet you believe is causing the issue.
- [X] Please [format your issue](https://quarto.org/bug-reports.html#formatting-make-githubs-markdown-work-for-us) so it is easier for us to read the bug report.
- [X] Please document the operating system you're running. If on Linux, please provide the specific distribution.

**Comment by cderv** [comment]
From what I see, your test not outputing as you expect are all related to Block like input and this is by design that we transform Block to Inline

https://github.com/quarto-dev/quarto-cli/blob/45018613fa50746c9a038427f76993999ee7a09f/src/resources/filters/quarto-pre/shortcodes-handlers.lua#L106-L113

So passing a markdown list or a code block won't work as you expect. Our doc is probably not enough clear about that as it says 
> Variables can be either simple values or can include **arbitrary markdown content**.

If you need to insert complex markdown content, you can leverage [includes shortcode](https://quarto.org/docs/authoring/includes.html). And for customization, you can probably use (or combine) [project profile content](https://quarto.org/docs/projects/profiles.html#profile-content). Project profile would allow you to insert content based on a profile setting. 

@cscheid @dragonstyle `var` shortcode insertion is meant for inlines values right ? I don't think we plan to change that considering `includes`. Asking because maybe I am not aware of possible change about this.

**Comment by tverbeiren** [comment]
That's what I expected, looking at how the bullet points where rendered.

Personally, I would like to use both includes/profiles as well as variables. The first allows me to create content about _topics_, the second would be used for project-specific customizations of those topics. In other words, I see them as complementary. Having the possibility to add markdown blocks as a variables makes a lot of sense in this case.

Follow-up question: what would break if we changed the above `function processValue` to:

```lua
 function processValue(val, name, t)     
   if type(val) == "table" then 
     if #val == 0 then 
       return { pandoc.Str( "") } 
     else
       return val
   ...
```

**Comment by dragonstyle** [comment]
I would also defer to @cscheid on this, but I know when we originally implemented support for `var`, a common use case was processing inline content where block content wouldn't make sense or be acceptable. It does look to me like perhaps we could know enough about the context to only force `var` values to inlines when we're in the inline context - something we could consider for a future release.

**Comment by tverbeiren** [comment]
That sounds great! Thanks for looking into this.

**Comment by cscheid** [comment]
@tverbeiren Just FYI, I'm going to edit your issue title to reflect the work that will go on this.

**Comment by cderv** [comment]
For reference, possibly related and duplicates of 

* https://github.com/quarto-dev/quarto-cli/issues/1644

**Comment by cscheid** [comment]
#6768 happens for a slightly different reason; it's not about the syntactic context of the shortcode invocation, but rather about the place in Quarto's pipeline (`include` is resolved in typescript ahead of the engines; most other shortcodes are resolved in Lua.)

**Comment by cscheid** [comment]
We implemented this in Quarto 1.5; shortcode handle can respond to context through the fifth parameter in the function call. See https://github.com/quarto-dev/quarto-web/pull/1275 (which will eventually be live on https://quarto.org/docs/extensions/shortcodes.html)
